odoo/addons/base/models/res_users.py

```diff
@@ -947,24 +947,20 @@ def create(self, vals_list):
             if 'groups_id' in values:
                 # complete 'groups_id' with implied groups
                 user = self.new(values)
-                group_public = self.env.ref('base.group_public', raise_if_not_found=False)
-                group_portal = self.env.ref('base.group_portal', raise_if_not_found=False)
-                if group_public and group_public in user.groups_id:
-                    gs = self.env.ref('base.group_public') | self.env.ref('base.group_public').trans_implied_ids
-                elif group_portal and group_portal in user.groups_id:
-                    gs = self.env.ref('base.group_portal') | self.env.ref('base.group_portal').trans_implied_ids
-                else:
-                    gs = user.groups_id | user.groups_id.mapped('trans_implied_ids')
+                gs = user.groups_id | user.groups_id.mapped('trans_implied_ids')
                 values['groups_id'] = type(self).groups_id.convert_to_write(gs, user.groups_id)
         return super(UsersImplied, self).create(vals_list)
 
     @api.multi
     def write(self, values):
+        users_before = self.filtered(lambda u: u.has_group('base.group_user'))
         res = super(UsersImplied, self).write(values)
         if values.get('groups_id'):
             # add implied groups for all users
             for user in self.with_context({}):
-                if not user.has_group('base.group_user'):
+                if not user.has_group('base.group_user') and user in users_before:
```

Comment:
I don't really get this.

Reply:
When we "demote" a user (make the user a portal user, typically) other groups are hidden in the views (groups like "salesman", etc.).
Obviously a portal user should not have this group; so in that case we remove all groups to make sure that the user does not have unexpected privileges.
It's not perfect because the user will thus not have the same groups that a normal portal user would have (display groups typically) but security-wise it is "safe".
Note that at least via the interface, the user type is a selection (and 'implied_ids' are only additive, they never remove anything), so the only way to remove the user group on a user is to explicitly remove it.